### PR TITLE
feat(storage): Restart delete resumable upload config new

### PR DESCRIPTION
**Pull request by shubhangi-google**
Feature to support Restart and Delete resumable upload functionality
```Ruby
@example   

storage = Google::Cloud::Storage.new
bucket = storage.bucket bucket_name
## sample upload_id=  "AKDAyIs3wDvKFtl2e4PopZi-YtGZDxFr2vlR4Z58T6hL9Seypi5W2s7vrRe6-Qw5Fur79g11bT7MtcsSu-vYRkm5nVUKPpm_ciLPn65GQDqNMQ"
#Initiating Resumable upload 
bucket.create_file file, file_name
#Restarting  Resumable upload 
bucket.restart_resumable_upload file, upload_id
#Deleting Resumable upload 
bucket.delete_resumable_upload upload_id
```
Implementation linked with - https://github.com/googleapis/google-api-ruby-client/pull/21896, https://github.com/googleapis/google-api-ruby-client/pull/23376
this PR is created as https://github.com/googleapis/google-cloud-ruby/pull/29215 seems to be corrupted

### CODE REVIEW

File: google-cloud-storage/lib/google/cloud/storage/bucket.rb [hunk 1]
```diff
@@ -1410,6 +1433,23 @@ def delete if_metageneration_match: nil, if_metageneration_not_match: nil
...
+        # Delete resumable upload
+        # @param [String] upload_id Unique Id of an resumable upload
+        #
+        # @example
+        #   require "google/cloud/storage"
+        #
+        #   storage = Google::Cloud::Storage.new
+        #
+        #   bucket = storage.bucket "my-bucket"
+        #   bucket.delete_resumable_upload file,upload_id
```
Comment (cpriti-os) [inline google-cloud-storage/lib/google/cloud/storage/bucket.rb]: This does not match the function signature below, should  it be 
```
# bucket.delete_resumable_upload upload_id
```

File: google-cloud-storage/lib/google/cloud/storage/bucket.rb [hunk 2]
```diff
@@ -717,6 +717,29 @@ def default_kms_key= new_default_kms_key
...
+        #
+        #   storage = Google::Cloud::Storage.new
+        #
+        #   bucket = storage.bucket "my-bucket"
+        #   bucket.restart_resumable_upload file,upload_id
+
+        def restart_resumable_upload file, upload_id
+          ensure_service!
+          ensure_io_or_file_exists! file
+          raise "Upload Id missing" unless upload_id
```
Comment (cpriti-os) [inline google-cloud-storage/lib/google/cloud/storage/bucket.rb]: Can we raise a specific error like ArgumentError maybe, here and below.

File: google-cloud-storage/test/google/cloud/storage/bucket_test.rb [hunk 3]
```diff
@@ -1408,6 +1408,36 @@
...
+
+    mock = Minitest::Mock.new
+    mock.expect :delete_resumable_upload, true,
+      [bucket.name, upload_id],
+      **delete_resumable_upload_args(options: {})
+    bucket.service.mocked_service = mock
+    bucket.delete_resumable_upload  upload_id
+    mock.verify
+  end
+
```
Comment (cpriti-os) [inline google-cloud-storage/test/google/cloud/storage/bucket_test.rb:1461]: Can we also add tests for failure cases like tests to verify that an ArgumentError is raised when upload_id is nil for both restart_resumable_upload and delete_resumable_upload.

File: google-cloud-storage/test/helper.rb [hunk 4]
```diff
@@ -391,6 +391,18 @@ def insert_object_args name: nil,
...
+    }
+  end
+
+  def restart_resumable_upload_args options: {}
+    {
+      options: options
+    }
+  end
+
   def get_object_args generation: nil,
```
Comment (cpriti-os) [inline google-cloud-storage/test/helper.rb:400]: Could these functions be refactored into a single function to avoid duplication? Feel free to mention if you have reasons to keep them separate.

File: google-cloud-storage/test/google/cloud/storage/bucket_test.rb [hunk 5]
```diff
@@ -1408,6 +1408,36 @@
     end
   end
 
+   it "restarts a resumable upload with upload_id" do
+    new_file_name = random_file_path
+    upload_id= "TEST_ID"
```
Comment (cpriti-os) [inline google-cloud-storage/test/google/cloud/storage/bucket_test.rb]: Nit: space around equal

File: google-cloud-storage/lib/google/cloud/storage/bucket.rb [hunk 6]
```diff
@@ -717,6 +717,29 @@ def default_kms_key= new_default_kms_key
...
+        #
+        # @example
+        #   require "google/cloud/storage"
+        #
+        #   storage = Google::Cloud::Storage.new
+        #
+        #   bucket = storage.bucket "my-bucket"
+        #   bucket.restart_resumable_upload file,upload_id
+
+        def restart_resumable_upload file, upload_id
```
Comment (cpriti-os) [inline google-cloud-storage/lib/google/cloud/storage/bucket.rb:739]: Just a thought: Does it align with GCS terminologies if we called this function  `resume_upload`, given its understandable  only resumable uploads can be resumes/restarted?

File: google-cloud-storage/lib/google/cloud/storage/bucket.rb [hunk 7]
```diff
@@ -1410,6 +1433,23 @@ def delete if_metageneration_match: nil, if_metageneration_not_match: nil
                                 user_project: user_project
         end
 
+        ##
+        # Delete resumable upload
```
Comment (cpriti-os) [inline google-cloud-storage/lib/google/cloud/storage/bucket.rb:1440]: It is helpful for method documentation to specify the return type. This guides users on what to expect from the method—whether it returns nil, a Boolean, an object, or raises exceptions. Here and above.

File: google-cloud-storage/lib/google/cloud/storage/bucket.rb (hunk 6 above)
Comment (cpriti-os) [inline google-cloud-storage/lib/google/cloud/storage/bucket.rb:739]: Could you clarify why we need the file argument, why isn't it possible to take the upload id and get all information related to the upload from it?

File: google-cloud-storage/test/google/cloud/storage/bucket_test.rb [hunk 8]
```diff
@@ -1408,6 +1408,36 @@
     end
   end
 
+   it "restarts a resumable upload with upload_id" do
```
Comment (cpriti-os) [inline google-cloud-storage/test/google/cloud/storage/bucket_test.rb]: It would be better if the return values from the new functions are verified to be as expected. Here and the test below.

File: google-cloud-storage/lib/google/cloud/storage/bucket.rb (hunk 6 above)
Comment (bajajneha27) [reply google-cloud-storage/lib/google/cloud/storage/bucket.rb:739]: @cpriti-os Other uploads can be resumed as well.

Comment (bajajneha27) [reply google-cloud-storage/lib/google/cloud/storage/bucket.rb:739]: File contains the data you need to upload.

File: google-cloud-storage/lib/google/cloud/storage/bucket.rb (hunk 7 above)
Comment (shubhangi-google) [reply google-cloud-storage/lib/google/cloud/storage/bucket.rb:1440]: added

**Comment by shubhangi-google** [comment]
closing this PR as somehow it includes commit from main branch creating a new PR- https://github.com/googleapis/google-cloud-ruby/pull/30882
